### DISCUSSION TITLE
Automatically refresh upon changes (build & watch with nodemon)

### DISCUSSION THREAD

**Discussion by yakirbu** [loopbackio/loopback-next#7154] [Ideas]
Hi,

If I would like loopback to automatically refresh after changes, how can I do that?
running 'nodemon start' doesn't seem to work, it runs the server, but doesn't restart upon changes.

(.P.S, is there a substitute to 'lb property', for loopback 4, to add a new property to a model?)

Thanks.

**Comment by dougal83** [comment]
This experimental setup might do it:
https://github.com/mightytyphoon/lb4-ng-quickstart

**Comment by jannyHou** [comment]
@dougal83 Thank you for providing the example repository combines LoopBack 4 and Angular. 
cc @dhmlau shall we create a page in our document to promote community example repositories like this ^ ?

**Comment by gregra81** [comment]
@yakirbu Use this https://www.npmjs.com/package/tsc-watch
And add this line to your `scripts` in package.json:
`"start:watch": "tsc-watch -b --onSuccess \"node .\"",`

**Comment by dhmlau** [comment]
Related to: https://github.com/strongloop/loopback-next/issues/2107

**Comment by dhmlau** [comment]
@gregra81, thanks for your information.  I've tried your suggestion above, however, I got an error saying: `error TS6369: Option '--build' must be the first command line argument.`.  I need to remove `-b` in order to get it run.  Did I miss something?  

@jannyHou, it's a good idea to get the inventory of the examples.  @bschrammIBM suggested in https://github.com/strongloop/loopback-next/issues/2087 that we should make the distinction between examples and tutorials, i.e. 
- tutorials: include instructions on how to build such an app
- examples: only shows how certain things can be done, no need for instructions.  

To expand that work, I think we can create a section there.  
@bschrammIBM, what do you think?

**Comment by psaunders** [comment]
I have gotten quite a long way down the nodemon + ts-node route.

If you go down this route you will run into a few thing, first you have to create a point of entry file for ts-node because index.ts is some sort of bundling file and not a Typescript version of index.js.

```ts
import * as application from './src';

module.exports = application;

const config = {
  rest: {
    port: +process.env.PORT || 3000,
    host: process.env.HOST || 'localhost',
    openApiSpec: {
      setServersFromRequest: true,
    },
  },
};

application.main(config).catch(err => {
  console.error('Cannot start the application.', err);
  process.exit(1);
});
```

Then your nodemon configuration to package.json
```json
  "nodemonConfig": {
    "watch": [
      "src"
    ],
    "ext": "ts",
    "exec": "ts-node nodemon.ts"
  },
```

Then you have to add the typescript controller files into your `application.ts`

```
    this.bootOptions = {
      controllers: {
        // Customize ControllerBooter Conventions here
        dirs: ['controllers'],
        extensions: ['.controller.js', '.controller.ts'],
        nested: true,
      },
    };
```

I now am getting a:  `The key repositories.XYZRepository was not bound to any value` and am getting the feeling that this is not worth it and just switched over to this:

```
  "nodemonConfig": {
    "watch": [
      "src"
    ],
    "ext": "ts",
    "exec": "npm start"
  },
```

Which is a bit slow but fine for now.

**Comment by bschrammIBM** [comment]
> This experimental setup might do it:
> https://github.com/mightytyphoon/lb4-ng-quickstart
@dougal83 Thank you for providing the example repository combines LoopBack 4 and Angular.
We would like to publish this example on our loopback.io doc site, in a new section entitled **Community Contributions**. Are you ok with having this example published? Please let me know!

**Comment by dougal83** [comment]
> > This experimental setup might do it:
> > https://github.com/mightytyphoon/lb4-ng-quickstart
> > @dougal83 Thank you for providing the example repository combines LoopBack 4 and Angular.
> > We would like to publish this example on our loopback.io doc site, in a new section entitled **Community Contributions**. Are you ok with having this example published? Please let me know!

It's not my code, perhaps you can ask @mightytyphoon who originally posted the example #1818

**Comment by bachdinhnhan** [comment]
> @yakirbu Use this https://www.npmjs.com/package/tsc-watch
> And add this line to your `scripts` in package.json:
> `"start:watch": "tsc-watch -b --onSuccess \"node .\"",`

@gregra81 
I started with your idea and the result is unexpected when it compiles entire project into the current directory instead of **/dist** folder. Eventually, it overrides the index.js file and makes the project unable to start with "npm run start". 

So after digging into **tsc-watch** and **tsc** itself, it ends up with a better version.
`"start:watch": "tsc-watch --target es2017 --outDir ./dist --onSuccess \"node .\"",`

**Comment by bachdinhnhan** [comment]
@dhmlau @yakirbu 
Try add this line to your scripts in package.json:
`"start:watch": "tsc-watch --target es2017 --outDir ./dist --onSuccess \"node .\"",`

It automatically detect any source-code changes and restart the server as well.

**Comment by aegyed91** [comment]
for the record: if u want to debug ur app and have proper sourcemap support. Use: `"start:watch": "tsc-watch --target es2017 --outDir ./dist --onSuccess \"node --inspect .\""` with [chrome canary](https://www.google.com/chrome/canary/) and [NiM](https://chrome.google.com/webstore/detail/nodejs-v8-inspector-manag/gnhhdgbaldcilmgcpfddgdbkhjohddkj?hl=en)

**Comment by tonifirnandes** [comment]
> for the record: if u want to debug ur app and have proper sourcemap support. Use: `"start:watch": "tsc-watch --target es2017 --outDir ./dist --onSuccess \"node --inspect .\""` with [chrome canary](https://www.google.com/chrome/canary/) and [NiM](https://chrome.google.com/webstore/detail/nodejs-v8-inspector-manag/gnhhdgbaldcilmgcpfddgdbkhjohddkj?hl=en)

Thanks, it works, but hopefully there is still another simple way, so our source code not fully of *.js.map file.

**Comment by pavelgronsky** [comment]
So, who can explain or who can show example about how I can used nodemon with Loopback 4. How I can changed files and take autorestart my loopback server?

**Comment by anjorinjnr** [comment]
Add
```
  "nodemonConfig": {
    "verbose": true,
    "watch": [
      "src/"
    ],
    "ignore": [
      "dist/*"
    ],
    "ext": "ts",
    "exec": "npm start"
  }
```
to your package.json

Then simply run nodemon (assuming you've already installed it)

**Comment by pavelgronsky** [comment]
@anjorinjnr What start file I need to run in nodemon command?
 **nodemon  index.ts**  - this is correct command?

**Comment by pavelgronsky** [comment]
Ok, work for me this option:
I added script to package.json

**"scripts": {
    "dev": "nodemon server.js"
  }**

and added this to package.json:

**"nodemonConfig": {
    "verbose": true,
    "watch": [
      "src/"
    ],
    "ignore": [
      "dist/*"
    ],
    "ext": "ts",
    "exec": "npm start"
  }**

And run application by this command:
**npm run dev**

**Comment by frbuceta** [comment]
> Ok, work for me this option:
> I added script to package.json
> 
> **"scripts": { "dev": "nodemon server.js" }**
> 
> and added this to package.json:
> 
> **"nodemonConfig": { "verbose": true, "watch": [ "src/" ], "ignore": [ "dist/*" ], "ext": "ts", "exec": "npm start" }**
> 
> And run application by this command:
> **npm run dev**

It works for me too. Thanks

**Comment by aegyed91** [comment]
> > for the record: if u want to debug ur app and have proper sourcemap support. Use: `"start:watch": "tsc-watch --target es2017 --outDir ./dist --onSuccess \"node --inspect .\""` with [chrome canary](https://www.google.com/chrome/canary/) and [NiM](https://chrome.google.com/webstore/detail/nodejs-v8-inspector-manag/gnhhdgbaldcilmgcpfddgdbkhjohddkj?hl=en)
> 
> Thanks, it works, but hopefully there is still another simple way, so our source code not fully of *.js.map file.

this solution works with `nodemon` + `lb-tsc --watch`. also modified `tsconfig.json` to generate inline sourcemaps which is also supported by normal chrome. no need to install canary chrome

modify package.json:
```json
"scripts": {
  "build:watch": "lb-tsc es2017 --outDir dist --watch",
  "start:dev": "nodemon --watch ./dist --inspect ./index.js"
}
```

modify tsconfig.json:
```json
{
  "$schema": "http://json.schemastore.org/tsconfig",
  "extends": "@loopback/build/config/tsconfig.common.json",
  "compilerOptions": {
    "rootDir": "src",
    "sourceMap": false,
    "inlineSourceMap": true,
    "inlineSources": true
  },
  "include": ["src"]
}
```

u will have to run both `npm run build:watch` and `npm run start:dev` in 2 separate shell

cc @tonifirnandes

**Comment by pavelgronsky** [comment]
> npm run dev
Ok, friends. 
I was able to understand how run nodemon by launch.json in VSCODE and how run nodemon by npm run dev command. Everything works fine, but how I can run nodemon together with debugged by npm command ?

**Comment by rodmcnew** [comment]
I think it would be best for adoption if loopback came with something that handled this out of the box. Installing the framework thinking you can start coding right away, then having to deal with this problem isn't a great experience.

- A new loopback app comes with 23 package.json scripts and none of them do this.
- Meanwhile create-react-app comes with 4 scripts and one does do this.

**Comment by Azzyxec** [comment]
I though of using Turbo repo along with the nodemon solution
it seems to have reduced the build times from 18s to 3s,

the basic idea is to run build watch and restart the server using nodemon

I have posted a [solution on stackoverflow](https://stackoverflow.com/q/74480889/2122822) so other can also give some feeback

Thanks hope its helps someone

**Comment by adrienrn** [comment]
I didn't want to introduce `nodemon` nor drift away from loopback default build; i.e. no custom TS, use `lb-tsc`, etc. Other people mentioned `tsc-watch` ( [here](https://github.com/loopbackio/loopback-next/discussions/7154#discussioncomment-465343), [here](https://github.com/loopbackio/loopback-next/discussions/7154#discussioncomment-465349)) and it has the advantage to be a one-liner solution and I've been happy with it in other projects:

```
$ yarn add -D tsc-watch
```

Then, add this to your `package.json`, and _done_:
```
"start:dev": "tsc-watch --noClear --compiler ./node_modules/.bin/lb-tsc --onSuccess 'node -r source-map-support/register .'",
```

Would be great to have that out-of-the-box 😬 !

**Reply by quicksilverr** [reply]
Hey any clue, about the cause of this  `error TS5042: Option 'project' cannot be mixed with source files on a command line.`
